include/common/dm.h

```diff
@@ -77,21 +82,78 @@ struct driver {
 };
 
 /**
- * Get the first device of a given class.
+ * Get the total number of subdevices (channels, etc.) available in a class.
+ *
+ * @param class One of the enumerated driver classes.
+ * @return      The number of available subdevices.
+ */
+uint8_t dm_count_subdevs_by_class(uint32_t class);
+
+/**
+ * Get the first device of a given class. This function only considers
+ * successfully-probed devices.
  *
  * @param class One of the enumerated driver classes.
+ * @return      The address of the device description, or NULL if no
+ *              devices of this class were found.
  */
-struct device *dm_get_by_class(uint32_t class);
+struct device *dm_first_dev_by_class(uint32_t class);
```

Comment:
This function is _highly_ problematic (it was before, also). If a code excerpt expects the first device returned to be the correct one due to alphabetical ordering, the addition of a new device with a higher alphabetical priority means that failing to change the dependent device results in unexpected behavior.

A possible alternative for this function is to return an array of devices that belong to the specified class. It can be accompanied by a function that returns the number of devices belonging to a class. This obviously leads to more code that requires maintenance due to the embedded nature of this project, but I personally believe changing a value associated with the number of devices belonging to a class in the device class enumeration is better than having to change existing code because the order of the devices has changed.

Reply:
The whole point of `dm_*_by_class` is you don't care which device or driver is returned, as long as something of the correct type is returned. For example, it doesn't matter which watchdog is used, as long as it works like a watchdog. For cases where you do care about the specific device you're interacting with, either it's specified as a clockdev/irqdev/etc. in `devices.c`, or you can use `dm_*_name`.

For the use case of, say, turning all of the clocks off. It doesn't matter which order the devices are returned in, since you're going to operate on all of them. For SCPI, there's a command to get the names of the clocks/sensors/regulator, etc., so the order shouldn't be important there either.